runtime/lua/vim/treesitter/query.lua

```diff
@@ -408,12 +455,24 @@ local directive_handlers = {
 
     -- If this produces an invalid range, we just skip it.
     if range[1] < range[3] or (range[1] == range[3] and range[2] <= range[4]) then
-      if not metadata[capture_id] then
-        metadata[capture_id] = {}
-      end
       metadata[capture_id].range = range
     end
   end,
+
+  -- Transform the content of the node
+  -- Example: (#gsub! @_node ".*%.(.*)" "%1")
+  ['gsub!'] = function(match, _, bufnr, pred, metadata)
```

Comment:
`gsub!` could also store it's result in metadata. This is how `#downcase!` in nvim-treesitter works. It would avoid an additional function and the need to change all usages of `get_node_text` in predicates. `#downcase!` is already used for injections where the language name is printed in uppercase.

Reply:
> gsub! could also store it's result in metadata

done

> It could be complicated to mutate meaning of the capture text as it is unclear whether #match? should work on the preprocessed or the original result.

currently predicates only work on the original result since `add_predicate`'s handler doesn't accept metadata, directives also need changes like https://github.com/neovim/neovim/commit/535ebbf66b644cc9ee06bc206bdf8340f765af63

Reply:
I was actually asking, why you want to transform the capture instead of giving the result a new name (=new metadata name unrelated to capture)

Reply:
This is for dynamic language injections, there is some explanation in the original pr

Reply:
No, there is no explanation in the original PR. You can dynamically inject a language with a `@language` capture but also with `.language` metadata. So why is a transformation of a capture necessary. `#down-case!` was also thought for dynamic language injection, it was first first modifying the capture and in the end we thought it would be easier to leave capture immutable. It didn't require changes to the original code and there was no ambiguity for plugins whether `@foo` would refer to the un-processed or processed version of it, which is more clear when you give input and output different names. 

Of course, magically transforming the capture text might be convenient as further predicates or plugins might automatically (automagically) profit from the transformed text. However, we did not have a clear answer in which order the predicates and directives run at that time and didn't dare to solve that problem at that time. When we're mutating capture, there should also be a clear documentation in which order the predicates/directives run to make clear whether they will see the transformed or untransformed text.

Reply:
> So why is a transformation of a capture necessary

Previously `#down-case!` was the only directive AFAIK that does something to `@language`, which is unintuitive to me since everything works for `@content`, and it took me a lot of debbuging to realize that they don't work as I was trying to implement dynamic injections for nix. This allows so that these directives do work, and makes it possible to implement injections where you need to do more than just downcasing like the one in the description with existing tools like `#offset!` and the added `#gsub!`. This was not previously possible AFAIK without implementing less intuitive directives like I did in https://github.com/nvim-treesitter/nvim-treesitter/pull/3902.

> there was no ambiguity for plugins whether @foo would refer to the un-processed or processed version of it

There was no ambiguity for `#downcase!`, but it was ambiguous to me and undocumented that all the other directives don't work as I expected.

> we did not have a clear answer in which order the predicates and directives run at that time and didn't dare to solve that problem at that time

`#downcase!` was the only directive that work for `@language`, and there was no other reasons for users to mix and match directives and predicates for `@language`. The behavior with predicates didn't change: they still operate on the unprocessed nodes, and I think it's a clear improvement that directives actually work even though they might be unordered - they used to not work at all. This should not break anything if it is implemented correctly.

Reply:
> Previously #down-case! was the only directive AFAIK that does something to @language, which is unintuitive to me since everything works for @content, and it took me a lot of debbuging to realize that they don't work as I was trying to implement dynamic injections for nix

I agree I thought we made it to be have like `#set!` with an explicit output but apparently I was wrong.